### ISSUE TITLE
Import ordering

### ISSUE THREAD

**Issue by kompot** [benmosher/eslint-plugin-import#31]
Are there any thoughts on implementing sorting rule? With logic like:

- first "global" libraries should be imported
- then local files (after \n\n)
- everything should be sorted

Does it make any sense for anyone?

**Comment by benmosher** [comment]
I can imagine breaking this up into a few rules, somewhat loosely based on Python's [style guide import conventions][pep8]:

- `import-first`: all `import` statements should precede any non-import statements
- `absolute-first`: all absolute imports should precede any relative imports
- `sorted`: need clarity on what this would enforce

I'm not sure what sort function you would apply. Personally, I generally import packages with broadest scope first, with the most specific last; this generally follows `absolute-first` as stated above. But I'm not sure how I would codify that. Only sort order I can imagine enforcing via linting is lexicographic order, but I wouldn't use it, personally. 

I'd certainly accept a PR for `sorted` and just leave it off by default, though. Sounds like it falls squarely in scope for the plugin.

[pep8]: http://legacy.python.org/dev/peps/pep-0008/#imports

**Comment by chase** [comment]
`import-first` and `absolute-first` sound like excellent additions. :+1:

**Comment by mikekreeki** [comment]
:+1:

**Comment by benmosher** [comment]
Alright, just published this with v0.7.9. Seems to work as expected for me, but you'll have to configure it in your `.eslintrc`. I have it turned off, by default.

I am not enforcing any spacing, just imports first and, if specified, absolutes before relatives. 

Still not doing `sorted` as there was never any clarification on what the comparator should be.

Closing for now, but comment back if it doesn't work as expected and I'll reopen as needed.

**Comment by gajus** [comment]
Can we re-open this to discuss sorting?

Specifically,

* alphabetical ordering based on the resource thats being imported
* ordering of aliases

> alphabetical ordering based on the resource thats being imported

```js
import 'aaa';
import 'bbb';
import 'ccc';
```

> ordering of aliases

```js
import {
    A,
    B,
    C
} 'abc';
```

**Comment by benmosher** [comment]
@gajus: can you create a new issue? "Rule proposal: import sorting" would be my druthers, but up to you.

**Comment by gajus** [comment]
> @gajus: can you create a new issue? "Rule proposal: import sorting" would be my druthers, but up to you.

* https://github.com/benmosher/eslint-plugin-import/issues/162
* https://github.com/benmosher/eslint-plugin-import/issues/163
